### PR TITLE
MOTECH-2665: Migrate admin settings

### PR DESCRIPTION
**Pull request by mkruszynski**

### CODE REVIEW

**Comment by nickdotreid** [comment]
@mkruszynski - sorry for the slow response to the review, here are a bunch of edits...

**(1)** information icon tool-tips are missing, I see you are trying to use the title property to fake this — but that's not accessible or styled (so its inconsistent across browsers) — use bs-popover (should already be included through [bootstrap-ui](https://angular-ui.github.io/bootstrap/#/popover))

Just btw — I am tempted to turn this pattern into a directive, as its repeated (everywhere) and ideally we wouldn't use something that only shows up on hover (or that we configure everywhere)

**(2)** Let's not follow the settings page's design directly -- you are bringing in some design elements and css that I should disappear... I'll make inline comments for these, but here is a brief list
- Top form fieldsets (upload settings, import/export, add settings location) should not have that weird float pattern, make them full width -- its easier to read and less markup
- Don't copy over the light blue text in dark blue header, its hard to read
- Don't use gradients, its fugly
- Primary action buttons (Save) should be aligned to the right, with width:auto
- You are using the legend element wrong, it must always be in a fieldset
- use [Sass variables](https://github.com/motech/motech-ui/blob/master/src/app/app.variables.scss) and don't use magic numbers in sass or css

File: src/admin/admin.routes.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    function adminRoutes ($stateProvider) {
+        $stateProvider
+        .state('admin', {
+            url: '/admin',
+            ncyBreadcrumb: {
+                label: 'Admin'
+            }
+        })
+        .state('admin.bundles', {
+            url: '/bundles',
```
Comment (nickdotreid) [inline src/admin/admin.routes.js]: It might make the most sense to leave the bundles implementation in their own file. I do think you made the correct change by prefixing each bundle route with 'admin.'

My issue is that the getBundle function at the bottom of the page makes the most sense within the directory with other bundles, or is an element that I should expose to the bundle controller directly.

File: src/admin/settings/settings.controller.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        $scope.label = function (key) {
+            return i18nService.getMessage('admin.settings.' + key);
+        };
+
+        $scope.saveSettings = function (settings) {
+            LoadingModal.open();
+            $http.post(ServerService.formatURL('module/admin/api/settings/platform'), settings)
+                .success( function() {
+                    LoadingModal.close();
+                    ModalWindow(angular.toJson(i18nService.getMessage('admin.settings.saved')),
```
Comment (nickdotreid) [inline src/admin/settings/settings.controller.js]: Is there a reason these calls to i18nService are wrapped in angular.toJson?

Since these are just success callbacks, it would make more sense to use [ModalAlert](https://github.com/motech/motech-ui/blob/master/src/common/modal/modal-alert.factory.js)

File: src/admin/settings/settings.controller.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            .error( function () {
+                LoadingModal.close();
+                ModalWindow(angular.toJson(i18nService.getMessage('admin.settings.error.export')),
+                    i18nService.getMessage('admin.settings.error')).open();
+            });
+        };
+
+        $scope.addValue = function(key, value) {
+            if ($scope.comboboxValues[key].indexOf(value) === -1) {
+                $scope.comboboxValues[key].push(value);
```
Comment (nickdotreid) [inline src/admin/settings/settings.controller.js]: This is out of scope of this ticket (slightly) - but the better UX for this field input would be for the option field to set to "" after the option is added, and the combo box to be set to the new option (so that the user can see something changed)

If you want to approach this functionality, please comment here — otherwise I will create a new ticket just for this item

File: src/admin/settings/settings.html [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                    </div>
+                    <span class="btn btn-primary" ng-click="uploadSettings()">{{'admin.submit' | translate}}</span>
+                </fieldset>
+            </form>
+        </div>
+    </div>
+</div>
+<div ng-if="!platformSettings.readOnly" class="form-group settings-box-small">
+    <div class="settings-box-content">
+        <legend class="frame-title settings-box-header">{{'admin.settings.location' | translate}}</legend>
```
Comment (nickdotreid) [inline src/admin/settings/settings.html]: The `<legend>` element only adds value to a page within a fieldset -- for every legend on this page, you should either change it to a heading element (h3 or h4) OR move it inside a fieldset (which is two lines below each item) ie:
```
<form>
    <fieldset>
        <legend>What the grouped fields do</legend>
        ...Add inputs and things....
   </fieldset>
</form>
```

File: src/admin/settings/settings.html [hunk 5]
```diff
@@ -0,0 +1,126 @@
+<div ng-if="!platformSettings.readOnly" class="form-group settings-box-medium">
+    <div class="settings-box-content">
```
Comment (nickdotreid) [inline src/admin/settings/settings.html]: Lets junk the settings-box-content classes (and divs) -- they add almost no value. I recommend turning them into forms or fieldsets within a larger form...

File: src/admin/settings/settings.html [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        <legend class="frame-title settings-box-header">{{'admin.settings.importexport.header' | translate}}</legend>
+        <form class="form-inline inside">
+            <fieldset class="form-content">
+                <label class="control-label input-label">{{'admin.settings.importexport.export' | translate}}</label>
+                <input type="button" ng-click="exportConfig()" value="{{'admin.settings.importexport.exportButton' | translate}}" class="btn btn-default"/>
+            </fieldset>
+        </form>
+    </div>
+</div>
+<div class="clearfix"></div>
```
Comment (nickdotreid) [inline src/admin/settings/settings.html]: Please get rid of the floated elements, and then remove this .clearfix element

File: src/admin/settings/settings.html [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+<div class="form-group">
+    <div class="settings-box-content" ng-show="platformSettings.settingsList.length > 0">
```
Comment (nickdotreid) [inline src/admin/settings/settings.html]: I don't think we need this ng-show element -- I'd remove the div and just leave the ng-repeat...

File: src/admin/settings/settings.scss [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    margin-top: 10px;
+    margin-bottom: 20px;
+    min-height: 20px;
+}
+.form-group{
+    >.settings-box-content{
+       border-radius: 4px;
+       border: 1px solid #D4E1F9;
+       padding: 0;
+       @include linear-gradient(to bottom, #eff8fd 0%, #F6FBFF 6%, #F6FBFF 16%, #F3FBFE 16%, #F0F9FE 37%, #F0F8FB 39%, #E8F5FB 57%, #E8F2FB 67%, #dbedf7 100%);
```
Comment (nickdotreid) [inline src/admin/settings/settings.scss]: Remove this gradient

File: src/admin/settings/settings.scss [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+       >.settings-box-header{
+           border-radius: 2px;
+           font-weight: bold;
+           line-height: 20px;
+           font-size: 15px;
```
Comment (nickdotreid) [inline src/admin/settings/settings.scss]: Use variables defined in /src/app/app.variables.scss

File: src/admin/settings/settings.scss [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+}
+.importexport{
+    float: left;
+    margin-top: 10px;
+    margin-bottom: 20px;
+    min-height: 20px;
+}
+.form-group{
+    >.settings-box-content{
+       border-radius: 4px;
```
Comment (nickdotreid) [inline src/admin/settings/settings.scss]: Don't set border-radius. I removed it for a reason

File: src/admin/settings/settings.scss [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+           padding: 5px 10px;
+           color: #c2eafe;
+           background: #144573;
+       }
+       >.settings-form{
+            border: transparent;
+            border-left:1px solid #D4E1F9;
+            padding: 10px 0 0 10px;
+            >legend{
+                color: #224477;
```
Comment (nickdotreid) [inline src/admin/settings/settings.scss]: Try not to make this level of a change to a legend -- we want styles across all pages to mostly look the same except for elements that are custom.

(I will make sure this is clear in the UI-Migration instructions, as it occurred to me that it might not be)

File: src/admin/settings/settings.scss [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+           margin: 10px;
+           input{
+               width: auto;
+               float: inherit;
+           }
+       }
+    }
+    .desc-label{
+        @extend .components;
+        width: 33.33333333%;
```
Comment (nickdotreid) [inline src/admin/settings/settings.scss]: (1) This is Sass, instead of writing a repeating value just, you can write `width: (1/3)%;` -- which will product an unrounded number.
(2) Don't style the interface this much

File: src/admin/settings/settings.scss [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+        width: 58.33333333%;
+        padding-right: 10px;
+    }
+    &.settings-box-small{
+        @extend .importexport;
+        width: 41.66666667%;
+        padding-left: 10px;
+    }
+    input[type="file"]{
+        position: absolute;
```
Comment (nickdotreid) [inline src/admin/settings/settings.scss]: This is too general a style to have a position:absolute on it -- why are we styling this so much?

File: src/admin/settings/settings.html [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+                                </select>
+                                <input class="input-medium form-control" ng-disabled="platformSettings.readOnly" type="text" ng-model="userValidator" />
+                                <a class="btn btn-success" ng-disabled="!userValidator" ng-click="addValue(option.key, userValidator)">
+                                    <span class="fa fa-fw fa-plus"></span>Add option</a>
+                            </div>
+                        </div>
+                    </div>
+                    <div class="form-group">
+                        <div class="btn-save">
+                            <input ng-if="!platformSettings.readOnly" type="submit" ng-click="saveSettings(pSettings)" value="{{'admin.save' | translate}}" class="btn btn-primary" />
```
Comment (nickdotreid) [inline src/admin/settings/settings.html]: What is up with the div around the input button? Please remove the div and just have the button, its more semantic.

**Comment by mkruszynski** [comment]
@nickdotreid Review this please

**Comment by sebbrudzinski** [comment]
@mkruszynski There are conflicts on this branch. Please resolve them before this is reviewed.

**Comment by mkruszynski** [comment]
@nickdotreid @sebbrudzinski Conflicts are resolved. Could you please take a look at this now.

**Comment by mkruszynski** [comment]
@nickdotreid Could you please review this.

**Comment by ngraczewski** [comment]
@nickdotreid 
Is there everything OK with this pull request? Is it ready do merge?

File: src/admin/settings/settings.html [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+<div ng-repeat="pSettings in platformSettings.settingsList">
+    <form id="platformSettingsForm">
+        <h4>{{'admin.settings.section.'+ pSettings.section | translate}}</h4>
+        <fieldset ng-repeat="option in pSettings.settings">
+            <div class="form-group">
+                <label for="{{option.key}}">
+                    {{'admin.settings.' + option.key | translate}}
+                    <span class="icon-documentation" data-trigger="hover" data-placement="left" bs-popover="{{'admin.settings.tooltip.' + option.key | translate}}" />
+                </label>
+                <input ng-if="isTextOrNumber(option)"ng-disabled="platformSettings.readOnly" type="text" id="{{option.key}}" name="{{option.key | translate}}" ng-model="option.value" />
```
Comment (jredlarski) [inline src/admin/settings/settings.html]: please add missing space after `ng-if="isTextOrNumber(option)"`

File: src/admin/settings/settings.html [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+                <input class="input-small" ng-if="isBool(option)" ng-disabled="platformSettings.readOnly" type="checkbox" id="{{option.key}}" name="{{option.key}}" ng-model="option.value" />
+                <div ng-if="isCombobox(option)" ng-init="addValue(option.key, option.value)">
+                    <select ng-model="option.value" ng-disabled="platformSettings.readOnly" ng-options="value for value in getComboboxValues(option.key)">
+                        <option ng-if="!option.value" value="">{{'mds.form.label.select' | translate}}</option>
+                    </select>
+                    <input ng-disabled="platformSettings.readOnly" type="text" ng-model="userValidator" />
+                    <a class="btn btn-success" ng-disabled="!userValidator" ng-click="addValue(option.key, userValidator)">
+                        <span class="fa fa-fw fa-plus"></span>Add option</a>
```
Comment (jredlarski) [inline src/admin/settings/settings.html]: change class to `icon-add` from [common files](https://github.com/motech/motech-ui/blob/master/src/common/base/base.icons.scss)

File: src/admin/settings/settings.html [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+            <div class="form-group">
+                <label for="new.db.user">{{'admin.label.db.username' | translate}}</label>
+                <input class="form-control input-large" id="new.db.user" type="text" name="db.username"/>
+            </div>
+            <div class="form-group">
+                <label for="new.db.password">{{'admin.label.db.password' | translate}}</label>
+                <input class="form-control input-large" id="new.db.password" type="text" name="db.host"/>
+            </div>
+        </fieldset>
+        <input type="submit" ng-click="saveNewSettings()" value="{{'admin.submit' | translate}}" class="btn btn-primary" />
```
Comment (jredlarski) [inline src/admin/settings/settings.html]: remove unnecessary class

File: src/admin/settings/settings.html [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+        </div>
+    </fieldset>
+    <input type="submit" value="{{'admin.submit' | translate}}"/>
+</form>
+<form>
+    <fieldset>
+        <h4>{{'admin.settings.importexport.header' | translate}}</h4>
+        <div class="form-group">
+            <label for="exportConfig">{{'admin.settings.importexport.export' | translate}}</label>
+            <input class="input-small" type="button" id="exportConfig" name="exportConfig" ng-click="exportConfig()" value="{{'admin.settings.importexport.exportButton' | translate}}" class="btn btn-default"/>
```
Comment (jredlarski) [inline src/admin/settings/settings.html]: change `class="button"` from [common](https://github.com/motech/motech-ui/blob/master/src/common/base/base.buttons.scss)

File: src/admin/settings/settings.html [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+                </label>
+                <input ng-if="isTextOrNumber(option)"ng-disabled="platformSettings.readOnly" type="text" id="{{option.key}}" name="{{option.key | translate}}" ng-model="option.value" />
+                <input class="input-small" ng-if="isBool(option)" ng-disabled="platformSettings.readOnly" type="checkbox" id="{{option.key}}" name="{{option.key}}" ng-model="option.value" />
+                <div ng-if="isCombobox(option)" ng-init="addValue(option.key, option.value)">
+                    <select ng-model="option.value" ng-disabled="platformSettings.readOnly" ng-options="value for value in getComboboxValues(option.key)">
+                        <option ng-if="!option.value" value="">{{'mds.form.label.select' | translate}}</option>
+                    </select>
+                    <input ng-disabled="platformSettings.readOnly" type="text" ng-model="userValidator" />
+                    <a class="btn btn-success" ng-disabled="!userValidator" ng-click="addValue(option.key, userValidator)">
```
Comment (jredlarski) [inline src/admin/settings/settings.html]: couldn't it simply be a `<button></button>` element with a class from [common files](https://github.com/motech/motech-ui/blob/master/src/common/base/base.buttons.scss)?

**Comment by jredlarski** [comment]
@mkruszynski After checking your changes I've noticed a few things:

**(1)** Try to use classes from common directory for styling buttons, icons etc. I've made inline comments for them.
**(2)** I can see that you have added bs-popover for icon tooltips as @nickdotreid requested but they don't seem to work.
**(3)** On chrome checkboxes look a bit strange, for example in **Require e-mail** field.

**Comment by jredlarski** [comment]
@mkruszynski Please resolve conflicts in this PR

**Comment by mkruszynski** [comment]
@jredlarski Done.

**Comment by jredlarski** [comment]
@nickdotreid 
I believe that @mkruszynski fixed all of our comments and it would be ready for merging, but I have problem with the way settings page looks...

Currently it looks like this:
motech-settings-top
motech-settings-bottom

I've seen that you requested to not follow exactly the same styling as we have in MOTECH and I agree with that. My problem is that sections like Security etc. are all combined together and it looks messy.

**Question**
Could you tell us what's your point of view or how you have imagined the look of the settings page?

File: src/admin/settings/settings.html [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+<div ng-repeat="pSettings in platformSettings.settingsList">
+    <form id="platformSettingsForm">
+        <h4>{{'admin.settings.section.'+ pSettings.section | translate}}</h4>
+        <fieldset ng-class="{'form-group': isBool(option)}" ng-repeat="option in pSettings.settings">
+            <legend ng-if="isBool(option)">
+                {{'admin.settings.' + option.key | translate}}
+                <span class="icon-documentation" data-trigger="hover" data-placement="left" bs-popover="{{'admin.settings.tooltip.' + option.key | translate}}" />
+            </legend>
+            <label ng-if="isBool(option)" class="checkbox">
+                <input ng-disabled="platformSettings.readOnly" type="checkbox" id="{{option.key}}" name="{{option.key}}" ng-model="option.value" />
```
Comment (nickdotreid) [inline src/admin/settings/settings.html]: The text for a checkbox should be to the right of the checkbox, and inside the label. You can remove the legend for boolean settings (as there is only one option).

File: src/admin/settings/settings.html [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+            <input class="button button-export" type="button" id="exportConfig" name="exportConfig" ng-click="exportConfig()" value="{{'admin.settings.importexport.exportButton' | translate}}" />
+        </div>
+    </fieldset>
+</form>
+
+<h3>{{'admin.settings.platform' | translate}}</h3>
+<div ng-repeat="pSettings in platformSettings.settingsList">
+    <form id="platformSettingsForm">
+        <h4>{{'admin.settings.section.'+ pSettings.section | translate}}</h4>
+        <fieldset ng-class="{'form-group': isBool(option)}" ng-repeat="option in pSettings.settings">
```
Comment (nickdotreid) [inline src/admin/settings/settings.html]: Is there a reason this isn't a `<div class="form-group">` -- a `<fieldset>` with one value doesn't make semantic sense.

**Comment by nickdotreid** [comment]
@jredlarski @mkruszynski 

I don't think its _entirely_ horrible either - the more I look at the page though, I think the "Save All" button isn't actually needed.

Here are some more specific problems and possible ways to fix them

(1) Checkboxes should have their label text to the right, and inside the `<label>` (I made an inline comment about this)

(2) Label text pushes below the input height (or looks like it) -- we might want to consider using flexbox, or a float based approach with our css styles for `.form-group` and `label` -- this way the height of the label gets registered.

(3) The visual difference between an <h3> and a <h4> style is pretty pathetic. I'd make <h3> bigger or bold -- alternatively, I'd make the <h4> into a legend, and wrap that in a fieldset (with the associated fields)

Is this helpful?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by mkruszynski** [comment]
@jredlarski @nickdotreid 
Thanks for advices.
"Save All" button has been deleted and section titles are bigger now.

File: src/admin/settings/settings.html [hunk 22]
```diff
@@ -21,49 +21,45 @@
...
-                    <button class="button-add" type="button" ng-disabled="!userValidator" ng-click="addValue(option.key, userValidator)">Add option</button>
-                </div>
+            </fieldset>
+            <div ng-if="isCombobox(option)" ng-init="addValue(option.key, option.value)">
+                <select ng-model="option.value" ng-disabled="platformSettings.readOnly" ng-options="value for value in getComboboxValues(option.key)">
+                    <option ng-if="!option.value" value="">{{'mds.form.label.select' | translate}}</option>
+                </select>
+                <input ng-disabled="platformSettings.readOnly" type="text" ng-model="userValidator" />
+                <a class="btn btn-success" ng-disabled="!userValidator" ng-click="addValue(option.key, userValidator)">
+                    <span class="fa fa-fw fa-plus"></span>Add option</a>
```
Comment (jredlarski) [inline src/admin/settings/settings.html]: @mkruszynski Why did you change button back to `<a></a>` element?

**Comment by mkruszynski** [comment]
@jredlarski I was changing the whole div, to do this i paste old version. I probably forgot to change this button.

**Comment by jredlarski** [comment]
@mkruszynski Sure, now it looks fine.

@nickdotreid After changes Settings page looks like this:
motech-settings-top-new
motech-settings-bottom-new
which I find more pleasant to the users eye. I believe this PR is ready for merging...

**Comment by nickdotreid** [comment]
@mkruszynski This looks so much better

The tooltip pop overs don't work (bs-popover) // can we fix this before merging?

File: src/admin/settings/settings.html [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+        <div class="form-group">
+            <label for="exportConfig">{{'admin.settings.importexport.export' | translate}}</label>
+            <input class="button button-export" type="button" id="exportConfig" name="exportConfig" ng-click="exportConfig()" value="{{'admin.settings.importexport.exportButton' | translate}}" />
+        </div>
+    </fieldset>
+</form>
+
+<h4 class="section-title">{{'admin.settings.platform' | translate}}</h4>
+<div ng-repeat="pSettings in platformSettings.settingsList">
+    <form id="platformSettingsForm">
```
Comment (nickdotreid) [inline src/admin/settings/settings.html]: @mkruszynski Is there a reason the ng-repeat isn't on the form element? This is a super minor thing, but it would be nice to have as few div's as possible

File: src/admin/settings/settings.scss [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+    .button-file{
+        height: auto;
+    }
+}
+.section-title{
+    font-size: 18px;
+}
+.partial-title{
+    font-size: 16px;
+}
```
Comment (nickdotreid) [inline src/admin/settings/settings.scss]: Instead of adding .section-title and .partial-title classes, can we add these [styles to h3 & h4 elements directly? (over here)](https://github.com/motech/motech-ui/blob/master/src/common/base/base.scss#L37) and possibly use em or a variable so we don't have "magic numbers" in our CSS?

File: src/admin/settings/settings.scss [hunk 25]
```diff
@@ -0,0 +1,25 @@
+.form-group{
```
Comment (nickdotreid) [inline src/admin/settings/settings.scss]: @mkruszynski Either prefix this code so it just works in this settings app, or [add it to forms.scss](https://github.com/motech/motech-ui/blob/master/src/common/form/form.scss#L125) (where it might screw other things up, but we can catch those and shape up the codebase together)

**Comment by mkruszynski** [comment]
@nickdotreid @jredlarski 
I made a corrections, according to the recommendations. I found that uib-tooltip directive isn't working quite well. When popover is too big, it's blinking. Should I fix it or this will be in separate ticket?

**Comment by sebbrudzinski** [comment]
test this please

File: src/common/base/base.scss [hunk 26]
```diff
@@ -34,8 +34,13 @@ a:hover{
     color: $brand-primary;
 }
 
-h3, h4{
-    font-size: 1em;
+h3{
```
Comment (jredlarski) [inline src/common/base/base.scss]: @nickdotreid @martokarski I believe styles for h3 and h4 should also have a prefix so they apply only in settings as they are causing a bit of a mess in other places in UI

File: src/common/base/base.scss [hunk 27]
```diff
@@ -34,8 +34,13 @@ a:hover{
...
+    font-size: $space-size + 4;
```
Comment (jredlarski) [inline src/common/base/base.scss]: @nickdotreid @mkruszynski Still we have some magic numbers here. Maybe a better idea is to either use em or define a `$h4-size` and `$h3-size` variables in [app.variables.scss](https://github.com/motech/motech-ui/blob/master/src/app/app.variables.scss)?

File: src/admin/settings/settings.html [hunk 28]
```diff
@@ -1,49 +1,49 @@
...
-        <h4 class="partial-title">{{'admin.settings.section.'+ pSettings.section | translate}}</h4>
+<div>
+    <h3>{{'admin.settings.platform' | translate}}</h3>
+    <form class="settings-form" id="platformSettingsForm" ng-repeat="pSettings in platformSettings.settingsList">
+        <h4>{{'admin.settings.section.'+ pSettings.section | translate}}</h4>
         <div class="form-group" ng-repeat="option in pSettings.settings">
             <label ng-if="!isBool(option)" for="{{option.key}}">
                 {{'admin.settings.' + option.key | translate}}
-                <span class="icon-documentation" data-trigger="hover" data-placement="left" bs-popover="{{'admin.settings.tooltip.' + option.key | translate}}" />
+                <span class="icon-documentation" uib-tooltip="{{'admin.settings.tooltip.' + option.key | translate}}"/>
```
Comment (jredlarski) [inline src/admin/settings/settings.html]: @nickdotreid @mkruszynski The blinking tooltip can be fixed by adding `tooltip-placement="left"`. The tooltip would show on the left (same as in the Motech Core) and without any blinking.

**Comment by jredlarski** [comment]
@mkruszynski Looks good! I've got only one issue, could you replace logic which is connected to the settings file upload by using `motech-file-upload` directive from this [PR#56](https://github.com/motech/motech-ui/pull/56)? There was a moment when three different PR's were using different ways of dealing with file uploading and it caused a little mess. Code in above mentioned PR was written based on your solution to this problem so is should be easy to make a use of it. For an example how it is used, please check latest commits in [PR#55](https://github.com/motech/motech-ui/pull/55).

File: src/admin/settings/settings.html [hunk 29]
```diff
@@ -0,0 +1,83 @@
+<legend class="frame-title">{{'admin.platform.settings' | translate}}</legend>
+<form class="settings-form" id="settingsFileForm" enctype="multipart/form-data">
+    <h4>{{'admin.settings.upload' | translate}}</h4>
+    <div class="form-group fileinput fileinput-new" data-provides="fileInput">
```
Comment (jredlarski) [inline src/admin/settings/settings.html]: @mkruszynski change data-provides to `"fileinput"` - I think this is causing problems on the UI

**Comment by mkruszynski** [comment]
@jredlarski On Firefox the problem is still existing, I really don't know what's causing it.

**Comment by jredlarski** [comment]
test this please

**Comment by jredlarski** [comment]
@mkruszynski Could you please specify the problem?

**Comment by mkruszynski** [comment]
@jredlarski After add a `<motech-file-upload>` markup, the settings upload section is looking like this:

ff

I used it in similar way like in posted earlier by you PR's.

**Comment by jredlarski** [comment]
@mkruszynski So it is only an appearance issue, right?

**Comment by mkruszynski** [comment]
@jredlarski Yes, everything else seems to be working.

**Comment by kmadej** [comment]
please resolve conflicts

**Comment by jredlarski** [comment]
test this please

File: src/common/motech-file-upload/motech-file-upload.html [hunk 30]
```diff
@@ -7,4 +7,6 @@
     <span class="fileinput-exists">{{'server.button.changeFile' | translate}}</span>
     <input type="file" id="fileInput" name="file" class="input-file"/>
 </span>
-<button class="button-delete fileinput-exists" data-dismiss="fileinput">{{'server.button.remove' | translate}}</button>
\ No newline at end of file
+<div>
```
Comment (jredlarski) [inline src/common/motech-file-upload/motech-file-upload.html]: @tomaszzawadzki Instead of adding a `<div>` element, please change the width of the `<div class="button-file">` so that both buttons are next to each other and fit within the size of the other form elements.

**Comment by jredlarski** [comment]
test this please

File: src/admin/settings/settings.html [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+                    <option ng-if="!option.value" value="">{{'mds.form.label.select' | translate}}</option>
+                </select>
+                <input ng-disabled="platformSettings.readOnly" type="text" ng-model="userValidator" />
+                <button class="button-add" type="button" ng-disabled="!userValidator" ng-click="addValue(option.key, userValidator)">Add option</button>
+            </div>
+        </div>
+        <input ng-if="!platformSettings.readOnly" type="submit" ng-click="saveSettings(pSettings)" value="{{'admin.save' | translate}}" />
+    </form>
+</div>
+<div class="settings-form" ng-show="platformSettings.settingsList.length <= 0">
```
Comment (pgesek) [inline src/admin/settings/settings.html]: This part doesn't make sense, we can remove it
